scripts/monitor/visualization/resource_monitoring.py

```diff
@@ -0,0 +1,596 @@
+
+    def __init__(self, meta_runtime: pd.DataFrame, metrics: pd.DataFrame) -> None:
+        """
+        initialization step: bring meta, runtime, and metrics data into scope
+        :param meta_runtime: meta_runtime table
+        :param metrics: metrics table
+        """
+        self.meta_runtime = meta_runtime
+        self.metrics = metrics
+        self.instances += 1
+
+    def run(self) -> None:  # maybe do this as part of initialization
+        """
+        Do everything that needs to be done to set up the webpage, and then set up the webpage
+        :return: None
+        """
+        self.generate_plot_files()
+        self.generate_report_homepage()
+
+    def generate_report_homepage(self) -> None:
+        """
+        Set up webpage
+        :return: None
+        """
+
+        pass
+        # jekyll stuff
+
+    def generate_plot_files(self) -> None:
+        """
+        Create all files and put them where other functions can find them
+        :return: None
+        """
+        tasks = self.num_task_shards()
+        unsharded_tasks = dict([(k, v) for k, v in tasks.items() if v == 0])
+        sharded_tasks = dict([(k, v) for k, v in tasks.items() if v != 0])
+        # unsharded_tasks = dict(filter(lambda elem: elem[1] == 0, tasks.items()))  # pycharm false positive?
+        # sharded_tasks = dict(filter(lambda elem: elem[1] != 0, tasks.items()))
+        for task in unsharded_tasks:
+            self.plot_aggregate(task, -1)
+        for task in sharded_tasks:
+            self.save_bokeh(task)
+
+    def combine(self, *args):  # UNTESTED: NO IDEA IF THIS WORKS
+        """
+        Combines multiple class instances (with their workflow data)
+        Meant for class instances with the same workflow
+        :param args: other class instances to combine- for data on multiple workflows
+        :return: None
+        """
+        Frames = list(self.meta_runtime)
+        for class_instance in args:
+            if (class_instance.meta_runtime.columns == self.meta_runtime.columns)[0]:
+                #
+                # NOTE TO SELF: this is stupid. It doesn't even work. Just check the workflow IDs.
+                # But do it later, when you fix the function for real
+                # Combining the datasets like this won't really work, figure something else out later
+                #
+                Frames.append(class_instance.meta_runtime)
+            else:
+                raise Exception("Not the same workflow, data cannot be combined")
+        self.meta_runtime = pd.concat(Frames)
+
+    def num_task_shards(self) -> Dict[str, int]:
+        """
+        Returns dict with task names as key and shard count as value
+        :return: dict of task:shard_count
+        """
+
+        task_and_shard_idx = self.meta_runtime.loc[:, ['runtime_task_call_name', 'runtime_shard']]
+        res = dict()
+        for task in task_and_shard_idx['runtime_task_call_name'].unique():
```

Comment:
one trivial suggestion: since you know `task_and_shard_idx` has exactly 2 columns, using index is probably easier and less typing.

Reply:
I'm not sure what you mean
use task_and_shard_idx[0] instead of task_and_shard_idx['runtime_task_call_name']?